docs/rfcs/1011-non-blocking-invite-transport.md

```diff
@@ -0,0 +1,1238 @@
+- the user manually cancelled the attempt
+- the hashed nonce didn't match the provided nonce
+- the SAS code communicated to the user was invalid
+- the payload could not be deciphered
+- the payload could not be deserialized
+- the payload contained invalid information
+- the attempt has been automatically cancelled by a new `start_attempt` command
+
+It is also registered that the greeter cancelled the attempt, so that this information can be provided in subsequent `invitation_cancelled` replies, along with the corresponding reason and timestamp. This way, the front-end application has all the information to properly communicate to the user what happened during the attempt.
+
+Pseudo-code:
+
+```python
+def invite_greeter_cancel_attempt(
+    token: InvitationToken,
+    attempt: AttemptID,
+    reason: CancelledAttemptReason,
+) ->
+  InvitationNotFound | InvitationCompleted | InvitationCancelled
+  AttemptNotFound | AttemptNotJoined | AttemptCancelled
+  GreeterNotAllowed | OK:
+    # Perform checks
+    [...]
+    # Call `greeter_cancel_attempt(reason)`
+    [...]
+```
+
+Schema definition:
+
+```json5
+[
+    {
+        "major_versions": [
+            4
+        ],
+        "req": {
+            "cmd": "invite_greeter_cancel_attempt",
+            "fields": [
+                {
+                    "name": "token",
+                    "type": "InvitationToken"
+                },
+                {
+                    "name": "attempt",
+                    "type": "AttemptID"
+                },
+                {
+                    "name": "reason",
+                    "type": "CancelledAttemptReason"
+                }
+            ]
+        },
+        "reps": [
+            {
+                "status": "ok",
+                "fields": [
+                    {
+                        "name": "attempt",
+                        "type": "AttemptID"
+                    }
+                ]
+            },
+            {
+                // The invitation token doesn't correspond to any existing invitation
+                "status": "invitation_not_found"
+            },
+            {
+                // The invitation has already been completed
+                "status": "invitation_completed"
+            },
+            {
+                // The invitation has been cancelled
+                "status": "invitation_cancelled"
+            },
+            {
+                // The author is not part of the allowed greeters for this invitation
+                "status": "author_not_allowed"
+            },
+            {
+                // The attempt id doesn't correspond to any existing attempt
+                "status": "attempt_not_found"
+            },
+            {
+                // The author did not join the attempt
+                // This should not happen, since joining is required to get the attempt ID
+                "status": "attempt_not_joined"
```

Comment:
Given this should not happen, I would be more practical to just merge this error into `attempt_not_found`:
```json5
{
                // No attempt joined by the user has this attempt ID
                "status": "attempt_not_found"
}
```

Just like for the merge of `greeter_not_found`&`greeter_not_allowed` I was talking above, this merge simplify the code and prevent probing attack (which are more theoretical than anything given our IDs are UUID4 and the routes have authentication, but still ! ^^)

Reply:
As explained in the comment before, I don't think merging status simplify the code, quite the contrary.

Although in that case, an argument could be made that this specific case happens only in the case of a bug and should be handled as such. However, this is not exactly an internal bug to the server (the attempt id could be shared through some other mean) which is why I made it a separate status. I don't have a strong opinion about dealing with it as an internal bug though.

Also I don't really see how the comment about probing attack is relevant here, in this specific context.

Reply:
> Also I don't really see how the comment about probing attack is relevant here, in this specific context.

If the attempt ID is a UUID, then this status allows to probe for valid attempt ID. I mentioned it because it is often considered as a dangerous design, but in our case I don't thing it's relevant (so I only see it as "it makes me uncomfortable, even if it should be safe" ^^)

If the attempt ID is not a UUID, then the error means the peer has joined a new attempt, and we use the new attempt ID before having ourselves joined. In this case no probing is possible whatsoever ;-)

> an argument could be made that this specific case happens only in the case of a bug and should be handled as such. However, this is not exactly an internal bug to the server

This situation is triggered by client-provided data, hence it's not an internal bug from the server point of view.

> which is why I made it a separate status

My point for merging `attempt_not_found` and `attempt_not_joined` is that they follow the same pattern: at this point the client is supposed to know the attempt he can use.
For instance if we consider attempt ID is a counter, and in our special example the current valid attempt ID is 10:
- if the client picks attempt ID < 10 he will get a `attempt_already_cancelled` (that's fine given the client knows about those IDs)
- if the client picks attempt ID == 10 he will get a `attempt_not_joined`
- if the client picks attempt ID > 10 he will get a `attempt_not_found`
I don't see the point on providing a dedicated "congratulation, you picked the right number !" error instead of just considering the attempt the client hasn't joined are not visible

Reply:
Per @mmmarcos decision, we keep those errors separated